### ISSUE TITLE
Writing to hdf5 file slows down enourmously for many long keys in the same group

### ISSUE THREAD

**Issue by mariaangelapellegrino** [h5py/h5py#1055]
To assist reproducing bugs, please include the following:
 * Operating System: Ubuntu 16.04  4.4.0-119-generic
 * Python version:   3.5.2
 * Where Python was acquired: system
 * h5py version: 2.8.0
 * HDF5 version 1.10.2
 * The full traceback/stack trace shown: none

We are converting data which has keys mapped to vectors into hdf5 format. After some time the writing slows down.

After several experiments we found that this happens when many datasets are written to a single group, and when these datasets have a long name (the longer the name, the earlier the problem occurs).
In the beginning the writing goes rather fast (+- 20MB/s), but after some point, the writing speed slows down to about 10 KB/s. After that point, the speed does not seem to go up again.

The size of the data does not affect. It seems to be some sort of limit on the amount and length of the keys.

The issue is reproducible with the code below:

```
import numpy as np
import h5py
import hashlib

def writeManyDatasets():
	file = h5py.File("myfile.h5", 'w')
	for i in range(0, 500000):
		data = np.asarray([1.0], dtype='float64')
		encodedName = hashlib.sha1(str(i).encode('utf-8')).hexdigest()
		encodedName = encodedName * 6
		#print (encodedName)
		dataset = file.create_dataset(encodedName, data=data)
		if i % 10000 == 0:
			print("Done with " + str(i))			
	file.close ()
```

**Comment by tacaswell** [comment]
This is a limitation of the way that hdf5 internally stores information about where the groups / datasets are in the file (iirc it is a B-tree of some sort). 

If you are using hashes as your names anyway, I suggest splitting the same up by groups 

```python
import numpy as np
import h5py
import hashlib
import time
from itertools import zip_longest

def writeManyDatasets():
        start = time.monotonic()
        with h5py.File("/tmp/myfile2.h5", 'w') as file:
                for i in range(0, 500000):
                        data = np.asarray([1.0], dtype='float64')
                        encodedName = hashlib.sha1(str(i).encode('utf-8')).hexdigest()
                        encodedName = encodedName * 6
                        #print (encodedName)
                        dataset = file.create_dataset(encodedName, data=data)
                        if i % 10000 == 0:
                                now = time.monotonic()
                                print(f"Done with {i} in {now - start}")
                                start = now
                                
def grouper(iterable, n, fillvalue=''):
        "Collect data into fixed-length chunks or blocks"
        # grouper('ABCDEFG', 3, 'x') --> ABC DEF Gxx"
        args = [iter(iterable)] * n
        return (''.join(n) for n in
                zip_longest(*args, fillvalue=fillvalue))

def writeManyDatasetsNested():
        start = time.monotonic()
        with h5py.File("/tmp/myfile2.h5", 'w') as file:
                for i in range(0, 500000):
                        data = np.asarray([1.0], dtype='float64')
                        encodedName = hashlib.sha1(str(i).encode('utf-8')).hexdigest()
                        encodedName = encodedName * 6
                        *groups, dset = grouper(encodedName, 50)
                        g = file
                        for group in groups:
                                g = g.require_group(group)
                        #print (encodedName)
                        dataset = g.create_dataset(dset, data=data)
                        if i % 1000 == 0:
                                now = time.monotonic()
                                print(f"Done with {i} in {now - start}")
                                start = now
 

```

These groups are still too big,

```
....
Done with 408000 in 0.46322193706873804
Done with 409000 in 0.49301140499301255
Done with 410000 in 0.46450832101982087
Done with 411000 in 0.47278940002433956
Done with 412000 in 5.889975662925281
Done with 413000 in 88.29981462203432
Done with 414000 in 88.23962989903521
Done with 415000 in 88.1731241439702

```

It is pretty amazing how much it hits a wall!



I'm going to close this as this isn't something h5py can help with, I suggest either doing something like the above spliting or contacting the hdf5 group.  It may also be worth densely packing all of your data into on large muliti dimensional array.

ps I have also had major issues with this and had to change file layout in the past.

**Comment by tkorach** [comment]
Same here - with over a million datasets in the root path, h5py becomes the bottleneck (tested with various disks, including fast SSD, all deteriorated to the same speed).

**Comment by aparamon** [comment]
@mariaangelapellegrino @tkorach Does opening the file with `libver='latest'` change the game? See [H5Pset_libver_bounds](https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetLibverBounds) for details. To my knowledge, that should be specifically helpful for enormous groups.

**Comment by tkorach** [comment]
@aparamon using `libver='latest'` did not change. Also, nesting datasets (instead of all datasets in the root group), so each group will contain fewer datasets did not help. 
I benchmarked various combinations (see code below), using representative data (to avoid bias to the compression step)

====================================================================
# compressed folder, no HDF5 compression per 99 batches of (50, 3, 25, 1024): average 0.021840, median 0.021840 seconds. Compression: 1.43GB to 1.16 GB
# HDF5 compression per 99 batches of (50, 3, 25, 1024): average 0.309534, median 0.309534 seconds. File: 713MB 
#  regular disk no HDF5 compression per 99 batches of (50, 3, 25, 1024): average 0.021060, median 0.021060 seconds files size 1.4 GB
# regular disk no HDF5 compression per 4999 batches of (50, 3, 25, 1024): average 0.045556, median 0.045556 seconds.periodic pauses in the middle, deteriorating from 0.02 to 0.045. file size 71.6GB
# Network drive No HDF5 compression per 99 batches of (50, 3, 25, 1024): average 0.066144, median 0.066144 seconds, size 1.43GB, took several minutes to flush
# Network drive HDF5 compression per 99 batches of (50, 3, 25, 1024): average 0.455374, median 0.455374 seconds size 713 MB
# compressed folder, NPZ: per 99 batches of (50, 3, 25, 1024): average 0.505440, median 0.505440 seconds 2.86 GB to 1.69 GB
# numpy compression per 99 batches of (50, 3, 25, 1024): average 0.413244, median 0.413244 seconds 560 MB
# ==================================================================
In conclusion, h5py compression is a bottleneck. Also h5py degrades with size, and relatively early.

**Comment by tacaswell** [comment]
@tkorach how have you separated this as as h5py issue vs an hdf5 issue?

I don't understand how compression is related to the issue of having too many names, could you please expand on that a bit?

**Comment by tkorach** [comment]
@tacaswell I did not separate between the two, since in Python I use HDF5 only through h5py and never directly. Thus, even if the problem is in h5py (and not the HDF5 library itself), it won't matter as I don't have any alternative wrapper. 
The number of names can interfere with HDF5 performance, the same way too many files in a single directory can slow down a regular file system. You can see measurements at :
https://stackoverflow.com/questions/35321093/limit-on-number-of-hdf5-datasets

**Comment by Michael-H777** [comment]
sorry to re-open this issue. 
Have there been any development in this problem? I have a similar situation. I am creating image patches for my deep learning model. Each patch is stored as a dataset. when working with large dataset, the number of dataset can easily to into 200k+. 
The write speed can drop to 20MB/s after im 20% in.  
Ubuntu 20, python 3.8, h5py 3.3 and writing into a SSD

**Comment by tacaswell** [comment]
@Michael-H777 Please re-ask this question in the hdf5 support forum.  The source of the bottle neck is in the underlying library.

**Comment by tkorach** [comment]
@Michael-H777 if all the datasets have the same shape, Numpy binary files are the fastest method. It does require manual book-keeping, though.
